demo-circuit/src/blaze_csw/constraints/mod.rs

```diff
@@ -0,0 +1,1680 @@
+            &should_enforce_wcert_hash,
+        )?;
+
+        // Enforce UTXO widthdrawal if required
+
+        if !csw_data_g.last_wcert_g.custom_fields_g.is_empty() {
+            // We use two custom fields (with half of the bits set) to store a single Field Element
+            assert_eq!(
+                csw_data_g.last_wcert_g.custom_fields_g.len(),
+                self.num_custom_fields as usize
+            );
+            assert!(self.num_custom_fields >= 2);
+
+            // Reconstruct scb_new_mst_root from firt 2 custom fields
+            let scb_new_mst_root = {
+                // Compute 2^128 in the field
+                let pow = FieldElement::one().double().pow(&[128u64]);
+
+                // Combine the two custom fields as custom_fields[0] + (2^128) * custom_fields[1]
+                // We assume here that the 2 FieldElements were originally truncated at the 128th bit .
+                // Note that the prover is able to find multiple custom_fields[0], custom_fields[1]
+                // leading to the same result but this will change the certificate hash, binded to
+                // the sys_data_hash public input, for which he would need to find a collision,
+                // and this is unfeasible.
+                let first_half = &csw_data_g.last_wcert_g.custom_fields_g[0];
+                let second_half = csw_data_g.last_wcert_g.custom_fields_g[1]
+                    .mul_by_constant(cs.ns(|| "2^128 * custom_fields[1]"), &pow)?;
+
+                first_half.add(
+                    cs.ns(|| "custom_fields[0] + (2^128) * custom_fields[1]"),
+                    &second_half,
+                )
+            }?;
+
+            csw_data_g
+                .utxo_data_g
+                .conditionally_enforce_utxo_withdrawal(
+                    cs.ns(|| "enforce utxo withdrawal"),
+                    &scb_new_mst_root,
+                    &csw_data_g.sys_data_g.nullifier_g,
+                    &csw_data_g.sys_data_g.amount_g,
+                    &should_enforce_utxo_withdrawal_g,
+                )?;
+        }
+
+        // Enforce FT withdrawal if required
+
+        csw_data_g.ft_data_g.conditionally_enforce_ft_withdrawal(
+            cs.ns(|| "conditionally enforce ft withdrawal"),
+            &sidechain_id_g,
+            self.range_size,
+            &csw_data_g.sys_data_g.mcb_sc_txs_com_end_g,
+            &csw_data_g.sys_data_g.nullifier_g,
+            &csw_data_g.sys_data_g.amount_g,
+            &should_enforce_ft_withdrawal_g,
+        )?;
+
+        // We check the public key ownership just once for both, choosing the appropriate public key
+        // and secret key, as it is an expensive check, we want to do it just once.
+        // NOTE: We could've done the same for nullifier and amount checks, but we didn't in order
+        //       to have cleaner code (we lose only 2 constraints anyway)
+
+        ScPublicKeyGadget::enforce_pk_ownership(
```

Comment:
In reality they can be both phantom or even not phantom, previously there was a check enforcing EXACTLY one of the two not being phantom (enforce_xor(should_enforce_utxo_withdrawal, should_enforce_ft_withdrawal). But after a discussion with Oleksandr we agreed that this constraint is useless as the prover cannot supply valid public inputs in either of the cases. If you have doubts about it we can reintroduce that constraint

Reply:
What do you mean they both (utxo and FT) can be phantom? Then how private key verification would pass?
My understanding is that private key verification ensures that one of them is not phantom. Isn't that correct?
Moreover, they can not both be not phantom and pass verification because nullifier cannot be the same for both UTXO and FT. So exactly one of them must be phantom to avoid conditional check. Where am I wrong?

Reply:
Yes you got it perfectly, that's why the XOR check would've been redundant

Reply:
Anyway we can add it, just to be on the safe side, it's just 1 constraint anyway

Reply:
I also do not see a strong reason to have this check, just wanted to confirm that I understood everything correctly.

Reply:
My five cents on this: Oleksandr's argument is perfeclty fine. However, a clarification of this in a comment would be helpful (it was giving me a hard time when reading the code).

Reply:
Added a comment about this